src/Text/Pandoc/Parsing.hs

```diff
@@ -1468,12 +1468,14 @@ citeKey = try $ do
   guard =<< notAfterString
   suppress_author <- option False (True <$ char '-')
   char '@'
-  firstChar <- alphaNum <|> char '_' <|> char '*' -- @* for wildcard in nocite
-  let regchar = satisfy (\c -> isAlphaNum c || c == '_')
-  let internal p = try $ p <* lookAhead regchar
-  rest <- many $ regchar <|> internal (oneOf ":.#$%&-+?<>~/") <|>
-                 try (oneOf ":/" <* lookAhead (char '/'))
-  let key = firstChar:rest
+  key <-
+    do firstChar <- alphaNum <|> char '_' <|> char '*' -- @* for wildcard in nocite
+       let regchar = satisfy (\c -> isAlphaNum c || c == '_')
+       let internal p = try $ p <* lookAhead regchar
+       rest <- many $ regchar <|> internal (oneOf ":.#$%&-+?<>~/") <|>
+                      try (oneOf ":/" <* lookAhead (char '/'))
+       pure $ firstChar:rest
+    <|> char '{' *> many (noneOf "}\\" <|> char '\\' *> anyChar) <* char '}'
```

Comment:
I wonder if this is too permissive. What about something like
```
@{hi there

- a list
- and another list item

another paragraph.}
```
Perhaps disallow newlines?
Do we want to allow spaces?

Reply:
Yes, that's a good point, could be confusing if one forgets the `}`. I think spaces are fine though.

Reply:
Another point here is that, if we disallow newlines they can still be included by escaping them, however I think it may be more useful to ignore escaped newlines, so that long identifiers can be split up over multiple lines. Kind of like Haskell Strings. In that case I would also want to skip (unescaped) whitespace after the escaped newline, to allow lining things up nicely.

```
- Item 1, @{a long \
  citekey}
- Item 2
```

Reply:
Re do we want to allow spaces: I think citekeys like `@{Hirshburg, 2016}` are useful.

Reply:
Also a related question, because I don't have the big overview of the parser. Is parsing ever supposed to fail? If we encounter an unescaped newline while parsing a citekey in braces, does parsing fail or should it just backtrack and treat the citekey as normal text and do we have to backtrack inside of the citekey parser or does the calling parser do that for us? Also right the citekey parser results in a committed failure if `\` is followed by eof. Is that correct behaviour?

Reply:
the whole citekey parser is under a `try`, so if anything fails inside it the whole thing will fail without consuming input. (and then, yes, the enclosing parser will likely parse this as regular text.)

Reply:
As for newlines, I think it would make sense to allow newlines only if not followed by a blankline (there's a parser for this).  And not to allow newlines to be backslash-escaped; that would create confusion with the existing markdown syntax for hard breaks.

Reply:
And these newlines should be interpreted as newlines or ignored? And if they should be ignored should leading whitespace after newlines also be ignored?

Reply:
The main consideration is that, if spaces are allowed, then in a hard-wrapped document one of the spacse might be turned into a newline + leading spaces. So, yes, I think it would be reasonable to parse a newline + leading spaces as a space for these purposes.  Not sure about backslash newlines.  Maybe it's cleanest just to allow them to be parsed as newlines.

Reply:
Ah, yes, that's a good consideration. I agree that parsing escaped newlines as newlines is the cleanest, however I'm not sure if it is that useful. On the other hand if we parse an unescaped newline as a space we still don't have a way to break up long citekeys without spaces into over multiple lines. We could instead ignore escaped newlines. I suppose the question is whether it's better to have useless behavior or non-intuitive, but also probably not harmful, rarely useful behavior.

Reply:
Also another thing would be to parse `\n` as a newline. That way it would actually be possible to use arbitrary citation keys. I don't really think anyone has citation keys with newlines inside them, but who knows.

Reply:
> we still don't have a way to break up long citekeys without spaces into over multiple lines

I think that's okay.  It's an issue that arises in many other contexts too (e.g. there's no way to split a long URL over multiple lines in a link).

As for newlines in citation keys: I'd prefer not to make that possible.